### PR TITLE
Support headers

### PR DESCRIPTION
**Pull request by braydensterrett**
### Summary
Summarize the contents of the code changes in your pull request. Tag any open issues you believe to be resolved by this pull request.

Adds the ability to specify custom headers to any request. 
Updates the logging framework to be more robust.
Requires 'route' keyword arguments to be provided for all request types. 'body' and 'header' keyword arguments may be provided, but have default values so they are not required. 

SCRIPT RUN:
screen shot 2018-05-18 at 4 07 08 pm

OUTPUT:
screen shot 2018-05-18 at 4 09 09 pm

### Additional Details
If you have anything else that you think may be relevant to this issue, list it here. Additional information can help us better understand your changes and speed up the review process.

Please add your name to the [CONTRIBUTORS.md] file. Adding your name to the [CONTRIBUTORS.md] file signifies agreement to all rights and reservations provided by the [License].

Thanks for contributing to icontrol_rest. 

[CONTRIBUTORS.md]: ../blob/master/CONTRIBUTORS.md
[License]: ../blob/master/LICENSE

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -41,12 +41,36 @@ api.get_sys_dns
       "search"=>["domain.com"]}
 
 # Or you can manually do it:
-api.get('/mgmt/tm/sys/dns')
+api.get(route: '/mgmt/tm/sys/dns')
```
Comment (JerrodCarpenter) [inline README.md]: Is this the breaking change? I don't want to introduce breaking changes unless it's required for new features.

File: lib/icontrol_rest/client.rb [hunk 2]
```diff
@@ -64,38 +64,51 @@ def logger
...
     end
 
     private
 
+    # Private - Builds a Hash of headers for a request.
+    #
+    # headers - A Hash of headers to be included with the required headers for an iControl request.
+    #
+    # Returns - A Hash of required headers and any additional headers to be included in the request.
+    def build_headers(headers:)
```
Comment (JerrodCarpenter) [inline lib/icontrol_rest/client.rb]: What's the benefit of moving the header structure into this method?

File: lib/icontrol_rest/client.rb [hunk 3]
```diff
@@ -64,38 +64,51 @@ def logger
     #         "nameServers"=>["1.2.3.72", "1.2.3.73"],
     #         "search"=>["domain.com"]}
     #
-    %w[delete get].each do |action|
-      define_method(action.to_s) do |route, _arg = nil|
-        send_request { self.class.send(action, route, @options) }
+    %w[delete get post put patch].each do |action|
```
Comment (JerrodCarpenter) [inline lib/icontrol_rest/client.rb]: Does this not duplicate what is below?

**Review by JerrodCarpenter — CHANGES_REQUESTED** [review]

File: lib/icontrol_rest/client.rb (hunk 3 above)
Comment (braydensterrett) [reply lib/icontrol_rest/client.rb]: Oh, darn. I tried to combine these but it didn't work out. I'll fix it.

File: lib/icontrol_rest/client.rb (hunk 2 above)
Comment (braydensterrett) [reply lib/icontrol_rest/client.rb]: Removed this. Was able to refactor a bit to keep this the way it was and still support a user to pass headers.

File: README.md (hunk 1 above)
Comment (braydensterrett) [reply README.md]: I've kept this portion passive, but have introduced the requirement of passing a `body` or `headers` keyword argument when one should be provided (if one or both of these should not be provided with a request, there are default values which will allow them not to be passed). This is non-passive, but completely required to support this new functionality.

File: lib/icontrol_rest/client.rb [hunk 4]
```diff
@@ -65,37 +64,40 @@ def logger
...
+    # body    - A Hash representing the request body to be passed with the request.
     #
     # Returns - response from the server as a hash.
     #
     # Raises  - RuntimeError when the server returns a response that isn't a 200
-    def send_request
+    def send_request(action:, route:, headers: {}, body: nil)
       Retriable.retriable on_retry: @retry_handler, tries: @tries do
-        response = yield
+        body = body.to_json unless body.nil?
```
Comment (JerrodCarpenter) [inline lib/icontrol_rest/client.rb:99]: if you set the default body to {}, then it'll allow you to send those empty bodies like iControl expects without this conditional logic.

Comment (braydensterrett) [reply lib/icontrol_rest/client.rb:99]: You would need more conditional logic because you can't pass an empty body to a `get` request lol It's asinine.

**Review by JerrodCarpenter — APPROVED** [review]

**Pull request merged**